### PR TITLE
Chore: TIP - Analytics Report Generator

### PR DESCRIPTION
**Pull request by ZarishIqbal**
### This PR contains the following:

- [x] Script to generate and upload teachers' activity analytics
- [x] Script to generate and upload TIP baseline schools' analytics

### CODE REVIEW

File: elixir/lib/sarkar/targeted_instructions/faculty.ex [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+              groups = Dynamic.get(peergroups_made, [level, subject])
+
+              groups
+              |> Enum.map(fn {_, quiz} ->
+                quiz
+              end)
+            end)
+          end)
+          |> Enum.flat_map(fn x -> x end)
+          |> Enum.flat_map(fn x -> x end)
```
Comment (mudssrali) [inline elixir/lib/sarkar/targeted_instructions/faculty.ex:285]: why we're doing flat map twice? We could use reduce to flat the list
```
Enum.reduce([], fn list, agg -> .... end)
```

File: elixir/lib/sarkar/targeted_instructions/faculty.ex [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+            |> Map.keys()
+
+          groups
+          |> Enum.map(fn quiz ->
+            "pg_#{quiz}"
+          end)
+        end)
+      end)
+      |> Enum.flat_map(fn x -> x end)
+      |> Enum.flat_map(fn x -> x end)
```
Comment (mudssrali) [inline elixir/lib/sarkar/targeted_instructions/faculty.ex:324]: again here

File: elixir/lib/sarkar/targeted_instructions/faculty.ex [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+            {:error, _error} ->
+              []
+          end
+
+        case Utils.start_school(school_id) do
+          {:ok, _pid} ->
+            IO.puts("#{index + 1} : #{school_id} started now")
+
+          _ ->
+            IO.puts("#{index + 1} : #{school_id} alread started")
```
Comment (mudssrali) [inline elixir/lib/sarkar/targeted_instructions/faculty.ex:364]: we don't need to have these log messages - these are a lot and repeatedly.

File: elixir/lib/sarkar/targeted_instructions/faculty.ex [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+
+  defp get_prod_json do
+    base_url = "https://storage.googleapis.com/targeted-instructions"
+
+    client =
+      Tesla.client([
+        {Tesla.Middleware.BaseUrl, base_url},
+        {Tesla.Middleware.JSON, engine: Poison}
+      ])
+
```
Comment (mudssrali) [inline elixir/lib/sarkar/targeted_instructions/faculty.ex:456]: we have this duplicate logic - can we move them to a reusable place?

File: elixir/lib/sarkar/targeted_instructions/grade_report_generator.ex [hunk 5]
```diff
@@ -1,49 +1,71 @@
...
+      {:ok, object} ->
+        IO.puts(object.mediaLink)
+
+      _ ->
+        "Not Uploaded O_o"
+    end
+
+    case Google.upload_csv("TIP_BASELINE_SCHOOLS_ANALYTICS_REPORT", school_csv) do
+      {:ok, object} ->
+        IO.puts(object.mediaLink)
```
Comment (mudssrali) [inline elixir/lib/sarkar/targeted_instructions/grade_report_generator.ex:43]: We can run them in task to achieve concurrency and optimization as
```ex
Task.start_link(fn ->
    grades_csv = TargetedInstructions.Schools.get_grades_report()
    case Google.upload_csv("TIP_BASELINE_SCHOOLS_REPORT", grades_csv) do
      {:ok, _object} ->
        IO.puts(object.mediaLink)
      _ ->
        "Not Uploaded O_o"
    end
end)
```

Same goes for other tasks.

**Review by mudssrali — CHANGES_REQUESTED** [review]
Just a few changes. Thanks @ZarishIqbal  - Also can you please documentation for newly added analytics CSV generation logic. This would help new developers to understand the logic easily.

**Comment by ZarishIqbal** [comment]
@mudssrali I have made the changes.

File: elixir/lib/sarkar/targeted_instructions/faculty.ex [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+              [school_id, facultyId, faculty_unique_id, date] ++ event_data
+            end)
+          end)
+          |> Enum.flat_map(fn x -> x end)
+
+        agg ++ result
+      end)
+
+    headers =
+      ~w(trainingactivity_wk1 trainingactivity_wk2 trainingactivity_wk4 techactivity_wk1 techactivity_wk2 techactivity_wk3 techactivity_wk4 techactivity_wk6 techactivity_wk7 techactivity_wk8)
```
Comment (mudssrali) [inline elixir/lib/sarkar/targeted_instructions/faculty.ex:428]: is that correct, `trainingactivity_wk1 trainingactivity_wk2 trainingactivity_wk4` ? we don't have training activity_wk3? @ZarishIqbal

Comment (ZarishIqbal) [reply elixir/lib/sarkar/targeted_instructions/faculty.ex:428]: If you check the csv, variable 3 is for youtube views which we left out purposely.

Comment (mudssrali) [reply elixir/lib/sarkar/targeted_instructions/faculty.ex:428]: Thanks @ZarishIqbal

**Pull request merged**